osgi/src/main/java/cucumber/java/runtime/osgi/OsgiClassFinder.java

```diff
@@ -25,38 +28,69 @@ public OsgiClassFinder(BundleContext bundleContext) {
+       
+    	final String searchPath = packageName.replace('.', '/');
 
-        final ArrayList<Class<? extends T>> result = new ArrayList<Class<? extends T>>();;
+        final ArrayList<Class<? extends T>> result = new ArrayList<Class<? extends T>>();
+        
         for (Bundle bundle : bundleContext.getBundles()) {
-            try {
-                result.addAll(findClassesInBundle(bundle, searchPath, parentType));
-            } catch (Exception e) {
-                LOGGER.error("Failed to inspect bundle " + bundle.getSymbolicName() + ": " + e.getMessage(), e);
-            }
+            
+        	/** 
+        	 * Fragments don't have a ClassLoader, so don't try to load classes from them.
+        	 * Their classes will be loaded inside their hosts. 
+        	 */
+        	if ((bundle.adapt(BundleRevision.class).getTypes() & BundleRevision.TYPE_FRAGMENT) == 0) {
+        	
+	        	try {
+	        		if (LOGGER.isDebugEnabled())
+	                    LOGGER.debug("Looking for sub classes of " + parentType.getName() + " in '" + packageName + "' package");
+	        		
+	                result.addAll(findClassesInBundle(bundle, "", searchPath, parentType));
+	                
+	                if (LOGGER.isDebugEnabled())
+	                    LOGGER.debug("Looking for fragment classes of " + parentType.getName() + " in '" + searchPath + "' path");
+	                
+	                result.addAll(findClassesInBundle(bundle, TARGET_CLASSES, TARGET_CLASSES+searchPath, parentType));
+	            } 
+	        	catch (Exception e) {
+	                LOGGER.error("Failed to inspect bundle " + bundle.getSymbolicName() + ": " + e.getMessage(), e);
+	            }
+        	}
         }
+        
         result.trimToSize();
+        
+        if (LOGGER.isDebugEnabled()) {
+        	for (Class<? extends T> clazz : result) {
+        		LOGGER.debug("Found class " + clazz.getName());
+        	}
+        }
+        
         return result;
     }
 
-    @Override
+    @SuppressWarnings("unchecked")
+	@Override
     public <T> Class<? extends T> loadClass(String className) throws ClassNotFoundException {
-        for (Bundle bundle : bundleContext.getBundles()) {
-            return (Class<? extends T>) bundle.loadClass(className);
+    	Class<? extends T> loadedClass = null;
+    	for (Bundle bundle : bundleContext.getBundles()) {
+            if ((loadedClass = (Class<? extends T>) bundle.loadClass(className)) != null) {
+            	break;
+            }
         }
+    	if (loadedClass != null) {
+    		return loadedClass;
+    	}
         throw new ClassNotFoundException("Couldn't load class from bundles: " + className);
     }
 
-    private <T> Collection<Class<? extends T>> findClassesInBundle(Bundle bundle, String searchPath, Class<T> parentType) {
+    protected <T> Collection<Class<? extends T>> findClassesInBundle(Bundle bundle, String prefix, String searchPath, Class<T> parentType) {
```

Comment:
I changed it to protected not for testing purposes (my tests only call the public method), but to allow the class to be extended for any future OSGi containers that may come along and need some slightly different functionality to discover classes.  In my actual project at work, I have a separate class that extends this, but because everything was private, I had to duplicate a ton of code in my new class.

Reply:
I changed it to protected not for testing purposes (my tests only call the public method), but to allow the class to be extended for any future OSGi containers that may come along and need some slightly different functionality to discover classes.  In my actual project at work, I have a separate class that extends this, but because everything was private, I had to duplicate a ton of code in my new class.

Reply:
Inheritance shouldn't be used to reuse code without also adhering to the LSP. It'd be better to use composition to solve this.